alien4cloud/alien4cloud-yorc-provider
# Automatic transformation of Docker-based topologies into Slurm/Singularity

# Pull Request description

## Description of the change

This PR adds a topology modifier that allow to transform a topology composed of docker jobs into a Slurm+Singularity topology. This is pretty much the equivalent of the Kubernetes modifier plugin that transform topologies composed of docker models into kubernetes applications.

This modifier supports only Jobs (deployments and statefulsets are not supported ATM).

### What I did

Transform `org.alien4cloud.extended.container.types.ContainerRuntime`, `org.alien4cloud.extended.container.types.ContainerJobUnit` and `tosca.nodes.Container.Application.DockerContainer` into a `yorc.nodes.slurm.SingularityJob` 

### How I did it

Everything is done within a single modifier: `docker-to-singularity-modifier`.
We added 3 new types in the Slurm location:   

* `yorc.nodes.slurm.ContainerRuntime` derived from `org.alien4cloud.extended.container.types.ContainerRuntime`
* `yorc.nodes.slurm.ContainerJobUnit` derived from `org.alien4cloud.extended.container.types.ContainerJobUnit`
* `yorc.nodes.slurm.HostToContainerVolume` derived from `org.alien4cloud.nodes.DockerExtVolume` this allows to mount a path from the host running the Singularity container into the container.

Those nodes are used only for matching. The modifier runs on `post_node_match` and transform those nodes into a `yorc.nodes.slurm.SingularityJob` 

### How to verify it

The following samples allows to demonstrate complex job orchestration that could be run on Kubernetes and Singularity/Slurm. https://github.com/ystia/tosca-samples/tree/develop/org/ystia/yorc/samples/job/text-count-demo

This sample is composed in 4 parts:
1. A `WebScraper` job downloads a file from an url and save it into a directory backed to an external volume. This volume should be matched against a  `yorc.nodes.slurm.HostToContainerVolume` and should be backed to a NFS directory on Slurm.
2. Then two jobs `LineCount̀` & `WordCount` are run in parallel they read the downloaded file from their volumes and compute the desired counts. Finally they output a resulting file.
3. Finally a `GCPBucketPublisher` job is run and copy the results files to a GCP Bucket.

The GCP bucket publishing may be tricky to implement as the GCP service account file should be provided into the `GCPSecret` volume. That mean that your service account file should be present on your nfs filesystem. This part can be removed or replaced by another simplest computation if needed.

Things to verify are:

* [ ] Jobs are properly sequenced

```
                               +------------------+
                               |                  |
                         +---->+   LineCount      +-----+
+------------------+     |     |                  |     |      +----------------------+
|                  |     |     -------------------+     |      |                      |
|   WebScraper     +-----+                              +----->+  GCPBucketPublisher  |
|                  |     |     +------------------+     |      |                      |
+------------------+     |     |                  |     |      +----------------------+
                         +---->+   WordCount      +-----+
                               |                  |
                               +------------------+
```
* [ ] Requested memory is propagated to Slurm (can be checked by `scontrol show job`). Memory limit is taken from the `mem_size` property of the `org.alien4cloud.extended.container.capabilities.ApplicationHost` capability on `org.alien4cloud.extended.container.types.ContainerRuntime`s 
* [ ] Requested CPU is propagated to Slurm (can be checked by `scontrol show job`). CPUs limit is taken from the `num_cpus` property of the `org.alien4cloud.extended.container.capabilities.ApplicationHost` capability on `org.alien4cloud.extended.container.types.ContainerRuntime`s
